### PR TITLE
fix for ON12 not allowing origin to be removed, even if no measures are attached

### PR DESCRIPTION
**Pull request by dougmills-DIT**
# TOPS-824 - cant remove the required origin from order number, which has no measures attached. 

The rule states : The quota order number origin cannot be deleted if it is used in a measure.

## Why
 * Rule would reject removal if any measures existed on the order, without checking specific measures for the geographical area

## What
* Updated rule, added test

Links to relevant material
See: https://uktrade.atlassian.net/browse/TOPS-824

### CODE REVIEW

File: quotas/business_rules.py [hunk 1]
```diff
@@ -163,11 +163,41 @@ class ON11(PreventDeleteIfInUse):
...
+        m_query = type(
+            on_origin.order_number.measure_set.first(),
+        ).objects.approved_up_to_transaction(
+            on_origin.transaction,
+        )
+
+        if not m_query.exists():
+            return
+
+        on_query = m_query.filter(
```
Comment (paulpepper-trade) [inline quotas/business_rules.py:187]: `order_numbers` in place of `on_query`?

**Review by paulpepper-trade — CHANGES_REQUESTED** [review]

File: quotas/business_rules.py [hunk 2]
```diff
@@ -163,11 +163,41 @@ class ON11(PreventDeleteIfInUse):
...
+
+        Get all the origins linked to this measure. Loop over these origins and
+        check that there are no measures linked to the origin .
+        """
+
+        m_query = type(
+            on_origin.order_number.measure_set.first(),
+        ).objects.approved_up_to_transaction(
+            on_origin.transaction,
+        )
```
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply quotas/business_rules.py:182]: There are issues referencing ORM objects in business rules. It raises an exception relating to partially loaded classes.  I could not see an example where direct calls to ORM classes are used in this file.

File: quotas/tests/test_business_rules.py [hunk 3]
```diff
@@ -231,6 +231,27 @@ def test_ON12(delete_record):
...
+
+    # add another origin
+    origin_without_measure = factories.QuotaOrderNumberOriginFactory.create(
+        order_number=measure.order_number,
+    )
+
+    # delete record
+    deleted = delete_record(origin_without_measure)
+
+    assert business_rules.ON12(deleted.transaction).validate(deleted) is None
```
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply quotas/tests/test_business_rules.py:252]: sure, can remove, wanted to be consistant about asserts in each test - but if its fine not to i can remove

File: quotas/business_rules.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply quotas/business_rules.py:182]: exact error : ImportError: cannot import name 'Measure' from partially initialized module 'measures.models' (most likely due to a circular import) (/Users/dougmills/PycharmProjects/tamato/measures/models.py)

File: quotas/business_rules.py [hunk 4]
```diff
@@ -163,11 +163,41 @@ class ON11(PreventDeleteIfInUse):
...
+            geographical_area_id=on_origin.geographical_area_id,
+            order_number_id=on_origin.order_number_id,
+        )
+
+        if on_query.exists():
+            raise self.violation(
+                model=on_origin,
+                message=(
+                    "The quota order number origin cannot be deleted if it is used in a"
+                    "measure."
```
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply quotas/business_rules.py:197]: good spot! done

File: quotas/business_rules.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply quotas/business_rules.py]: Good shout, thanks. Updated

File: quotas/business_rules.py [hunk 5]
```diff
@@ -163,11 +164,39 @@ class ON11(PreventDeleteIfInUse):
...
+            order_number_origin.transaction,
+        )
+
+        if not measures.exists():
+            return
+
+        order_numbers = measures.filter(
+            geographical_area_id=order_number_origin.geographical_area_id,
+            order_number_id=order_number_origin.order_number_id,
+        )
```
Comment (paulpepper-trade) [inline quotas/business_rules.py:189]: A comment around this section is showing as resolved, but the code looks unchanged with no response. I think the suggested change was this:
```suggestion
        order_numbers = measures.filter(
            geographical_area=order_number_origin.geographical_area,
            order_number=order_number_origin.order_number,
        )
```

**Comment by gabelton** [comment]
Does the rule only get called when an origin is being deleted? Should it have the `skip_when_not_deleted` decorator?

**Review by paulpepper-trade — CHANGES_REQUESTED** [review]

File: quotas/business_rules.py (hunk 5 above)
Comment (paulpepper-trade) [reply quotas/business_rules.py:189]: ... but Gabe's question about filtering by SID is also relevant.

File: quotas/business_rules.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulpepper-trade) [reply quotas/business_rules.py:196]: I don't see the change reflected in the source code, unless you've yet to push the change?

File: quotas/business_rules.py (hunk 5 above)
Comment (dougmills-DIT) [reply quotas/business_rules.py:189]: I am going to favour the SID update, since it has potential for a wider catchment

Comment (dougmills-DIT) [reply quotas/business_rules.py:189]: ok - getting to SID is erroring. Not sure why - falling back to object (which under the hood i expect goes on PK anyway)

Comment (dougmills-DIT) [reply quotas/business_rules.py:189]: ok worked it out - referencing SID working now

File: quotas/business_rules.py [hunk 6]
```diff
@@ -163,11 +164,39 @@ class ON11(PreventDeleteIfInUse):
...
+
+        measures = measures_models.Measure.objects.approved_up_to_transaction(
+            order_number_origin.transaction,
+        )
+
+        if not measures.exists():
+            return
+
+        order_numbers = measures.filter(
+            geographical_area_id=order_number_origin.geographical_area_id,
```
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply quotas/business_rules.py:187]: now using SID for query

File: quotas/business_rules.py (hunk 5 above)
Comment (dougmills-DIT) [reply quotas/business_rules.py]: @paulpepper-trade resolve if you are ok with this update ˆ

File: quotas/business_rules.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply quotas/business_rules.py]: its there - not sure why didn't show - i saw old version also

[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply quotas/business_rules.py]: i can see space now

[the remark this answers is not in the document]
Comment (paulpepper-trade) [reply quotas/business_rules.py]: Looks as though it's still missing the space after `" measure."`.

**Review by paulpepper-trade — APPROVED** [review]

**Review by gabelton — APPROVED** [review]

**Pull request merged**